File: light-control/algorithms/_meta/pixel.py
import colorsys

class Pixel:
    hue = 0.0
    sat = 0.0
    val = 0.0
    white = 0.0

    def __init__(self, hue=0, sat=0, val=0, white=0):
        self.set(hue, sat, val, white)

    def set(self, hue, sat, val, white=0):
        self.hue = hue
        self.sat = sat
        self.val = val
        self.white = white
    
    def clear(self):
        self.set(0, 0, 0, 0)

    def multiply(self, value):
        return Pixel(self.hue, self.sat, self.val * value, self.white * value);
    
    def diff(self, other):
        return Pixel(
            self.hue - other.hue,
            self.sat - other.sat,
            self.val - other.val,
            self.white - other.white
        )

    def combine_with(self, other_pixel):
        my_rgb = self.get_raw_rgb()
        other_rgb = other_pixel.get_raw_rgb()

        new_rgb = (
            min(my_rgb[0] + other_rgb[0], 1.0),
            min(my_rgb[1] + other_rgb[1], 1.0),
            min(my_rgb[2] + other_rgb[2], 1.0),
        )
        self.from_rgb(*new_rgb)

    def from_rgb(self, r, g, b):
        [hue, sat, val] = colorsys.rgb_to_hsv(r, g, b);
        self.hue = hue
        self.sat = sat
        self.val = val
    
    def get_raw_rgb(self):
        return colorsys.hsv_to_rgb(self.hue, self.sat, self.val)

    def get_rgb(self, brightness):
        max_val = 255
        rgb = colorsys.hsv_to_rgb(self.hue, self.sat, self.val * brightness)
        return (rgb[0] * max_val, rgb[1] * max_val, rgb[2] * max_val)
    
    def get_rgbw(self, brightness):
        rgb = self.get_rgb(brightness)
        max_val = 255 * brightness
        return (rgb[0], rgb[1], rgb[2], self.white * max_val)

File: lights-control-server/src/algorithms/_meta/pixel.ts
import { hsvToRgb } from './helpers'

export class Pixel {
  hue: number
  sat: number
  val: number
  white: number

  constructor(hue = 0, sat = 0, val = 0, white = 0) {
    this.hue = hue
    this.sat = sat
    this.val = val
    this.white = white
  }

  set(hue: number, sat: number, val: number, white = 0): void {
    this.hue = hue
    this.sat = sat
    this.val = val
    this.white = white
  }

  clear(): void {
    this.set(0, 0, 0, 0)
  }

  multiply(value: number): Pixel {
    return new Pixel(this.hue, this.sat, this.val * value, this.white * value)
  }

  diff(other: Pixel): Pixel {
    return new Pixel(
      this.hue - other.hue,
      this.sat - other.sat,
      this.val - other.val,
      this.white - other.white,
    )
  }

  combineWith(other: Pixel): void {
    const myRgb = this.getRawRgb()
    const otherRgb = other.getRawRgb()
    const r = Math.min(myRgb[0] + otherRgb[0], 1.0)
    const g = Math.min(myRgb[1] + otherRgb[1], 1.0)
    const b = Math.min(myRgb[2] + otherRgb[2], 1.0)
    this.fromRgb(r, g, b)
  }

  fromRgb(r: number, g: number, b: number): void {
    const max = Math.max(r, g, b)
    const min = Math.min(r, g, b)
    const delta = max - min

    let hue = 0
    if (delta !== 0) {
      if (max === r) hue = ((g - b) / delta) % 6
      else if (max === g) hue = (b - r) / delta + 2
      else hue = (r - g) / delta + 4
      hue /= 6
      if (hue < 0) hue += 1
    }

    this.hue = hue
    this.sat = max === 0 ? 0 : delta / max
    this.val = max
  }

  getRawRgb(): [number, number, number] {
    return hsvToRgb(this.hue, this.sat, this.val)
  }

  getRgb(brightness: number): [number, number, number] {
    const [r, g, b] = hsvToRgb(this.hue, this.sat, this.val * brightness)
    return [r * 255, g * 255, b * 255]
  }

  getRgbw(brightness: number): [number, number, number, number] {
    const [r, g, b] = this.getRgb(brightness)
    return [r, g, b, this.white * 255 * brightness]
  }
}
